File: python/two_pointers/611. Valid Triangle Number.py
from typing import List


class Solution:
    def triangleNumber(self, nums: List[int]) -> int:
        if len(nums) < 3:
            return 0
        nums.sort()
        result = 0
        for i in range(0, len(nums)):
            if nums[i] == 0:
                continue
            k = i + 2
            for j in range(i + 1, len(nums) - 1):
                while(k < len(nums) and nums[i] + nums[j] > nums[k]):
                    k += 1
                result += k - j - 1
        return result
    # def triangleNumber(nums):
    #     nums.sort()
    #
    # count = 0
    # for i in range(len(nums) - 1, 1, -1):
    #     left = 0
    #     right = i - 1
    #     while left < right:
    #         if nums[left] + nums[right] > nums[i]:
    #             count += right - left
    #             right -= 1
    #         else:
    #             left += 1
    #
    # return count

File: two-pointers/611. Valid Triangle Number.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var triangleNumber = function (nums) {
  nums.sort((a, b) => a - b);
  
  let result = 0;
  for (let i = 0; i < nums.length - 2; i++) {
    let k = i + 2;
    for (let j = i + 1; j < nums.length - 1 && nums[i] !== 0; j++) {
      while (k < nums.length && nums[i] + nums[j] > nums[k]) {
        k++;
      }
      result += k - j - 1;
    }
  }
  return result;
};

// const nums = [2, 2, 3, 4]; //3
// const nums = [2, 2, 3, 4, 5]; //6
const nums = [2,4,3,4]; // 4
// tslint:disable-next-line:no-console
console.log('triangleNumber: ', triangleNumber(nums)); // eslint-disable-line no-console
